[Feature] network model

yjjem opened the pull request:
## PR 정보 
- **Reviewer :** @1Consumption 
- **Feature :** `F1.1`, `F1.2`
- **Branch**: `feature/network-model` -> `develop`

## 진행 사항

### 기능
- HTTPMethod 타입 추가 3b4a776bfd6822a8d071b60205ccadbdadaa0572
- NetworkError 타입 추가 de74c798e20bfa57ed16073f80ed2dcdc6aca243
- Networkable( NetworkManager Interface ) 추가 2bc790ade2c26faa92676e4bc9f922f82646dc32
- NetworkManager 추가 fe957e88ad78fad457979d66344f277afc428508
    > load 및 upload 기능 구현

### 리펙토링
- 중복되는 세션 검증 코드 분리 69d141e1a95ca8ef7544f611f14e7f9766c899bc
    ```swift
  private func checkSessionValidity(error: Error?, response: URLResponse?) -> NetworkError? {
          
          if let error = error {
              return .requestFailed(error)
          }
          
          guard let response = response as? HTTPURLResponse else {
              return .notHTTPURLResponse
          }
          
          guard (200...299) ~= response.statusCode else {
              return .badResponse(response.statusCode)
          }
          
          return nil
     }
    ```

- 불필요한 session이 진행되지 않도록 수정했습니다. 847fb54b07997df41c1a96309af939e842a81b10
  ```swift
  deinit {
      session.invalidateAndCancel()
  }
  ```
    > View에서 벗어났음에도 불구하고 불필요한 session task가 진행되는 상황을 고려하여 struct -> class 로 전환하여 deinit시 invalidateAndCancel()이 불리도록 수정했습니다.

## 테스트
#### load
  - [x] test_서버_responseCode가_300일때_badResponse와_code를_올바르게_반환하는지
  - [x] test_request_생성_실패시_requestFailed_오류를_반환하는지
  - [x] test_response_200일시_success와_데이터를_반환하는지
#### upload
  - [x] test_성공적으로_upload시_에러반환을_안하는지
  - [x] test_upload_responseCode가_300일시_badResponse를_반환하는지
  - [x] test_upload_requestFaild시_에러를_올바르게_반환하는지

yjjem committed:
feat: HTTPMethod 추가

yjjem committed:
feat: NetworkError 추가

yjjem committed:
feat: Network Interface 추가

yjjem committed:
feat: NetworkManager 추가

yjjem committed:
feat: 중복되는 session validation 로직 메서드화 -> checkSessionValidity

yjjem committed:
refactor(NetworkManager): deinit시 session을 invalidate하도록 수정

yjjem committed:
chore: 코드 컨벤션 수정

1Consumption commented on remy-movie/remy-movie/Data/NetworkManager.swift:
```diff
@@ -0,0 +1,100 @@
+        return request
+    }
+    
+    private func checkSessionValidity(error: Error?, response: URLResponse?) -> NetworkError? {
+        
+        if let error = error {
+            return .requestFailed(error)
+        }
+        
+        guard let response = response as? HTTPURLResponse else {
+            return .notHTTPURLResponse
+        }
+        
+        guard (200...299) ~= response.statusCode else {
+            return .badResponse(response.statusCode)
+        }
+        
+        return nil
+    }
```
흠... 네트워크 응답이 유효한지를 체크하는 로직 같은데 nil을 반환했을 때가 유효한 응답인게 뭔가 이상해보입니다. 조금 더 자연스럽게 바꿔볼 수 있을까요?

yjjem replied:
확실히 말씀해주신대로 nil은 부자연스럽게 느껴집니다. Result 타입을 활용하여 아래와 같이 변경해보았습니다.
결과는 `failure`또는 `success`로 나뉘기 때문에 `tryValidateSession()`으로 메서드명을 수정하였고 성공시 response code를 반환하도록 변경했습니다. 

> 수정 커밋: 5d6f7ed5cccfb9c64a0a891e05e12777c1d716de

```swift
private func tryValidateSession(
        error: Error?,
        response: URLResponse?
    ) -> Result<Int, NetworkError> {
        
        if let error = error {
            return .failure(.requestFailed(error))
        }
        
        guard let response = response as? HTTPURLResponse else {
            return .failure(.notHTTPURLResponse)
        }
        
        guard (200...299) ~= response.statusCode else {
            return .failure(.badResponse(response.statusCode))
        }
        
        return .success(response.statusCode)
    }
``` 

실제로 사용하는 곳에서는 당장 success시 response code를 받는 곳이 없어 if-case-let 으로 failure만 처리했습니다.

```swift
let sessionValidation = self?.tryValidateSession(error: error, response: response)
            
    if case let .failure(validationFailed) = sessionValidation {
        completion(validationFailed)
        return
    }
```

1Consumption reviewed (commented):
고생하셨습니다.

피알이 워낙 작아 별 다른 코멘트는 없습니다.
다만 테스트 코드가 추가되면 더 좋겠습니다.

yjjem committed:
refactor(NetworkManager): session validation 로직 수정

yjjem committed:
docs: add NetwokrPlatformTest

yjjem committed:
test: add MockURLProtocol

yjjem committed:
test(load): badResponse 확인하는 테스트 추가

yjjem committed:
test(load): request fail 확인 테스트 추가

yjjem committed:
test(load): response 200일시 success와 함께 데이터 반환 확인 테스트 추가

yjjem committed:
test(MockURLProtocol): 불필요한 print문 제거

yjjem committed:
test: 200 response 테스트 XCTFail 추가

yjjem committed:
test(upload): 성공적으로 upload시 에러 반환 테스트 추가

yjjem committed:
test(upload): badResponse 300 반환 테스트 추가

yjjem committed:
test(upload): requestFail 에러 반환 테스트 추가

yjjem commented:
## PR 내용 추가
### 테스트를 위한 커스텀 URLProtocol 구현
- URLProtocol을 상속하여 테스트를 위한 MockURLProtocol을 구현했습니다.
- MockURLProtocol의 requestHandler를 통해 dummy response, data를 넣어주어 원하는 결과를 어느정도 유도할 수 있었습니다.
다만 아쉬운 점이 있다면 emptyData, notHTTPURLResponse등의 에러는 테스트해볼 수 없었습니다.

### 테스트 케이스 추가
#### load
- load 성공 -> 데이터 반환 케이스
- request 생성 실패 -> requestFailed 반환 케이스
- 서버 response 300 -> badResponse 반환 케이스
#### upload
- upload 성공 -> 에러 반환하지 않는 케이스
- request 생성 실패 -> requestFailed 반환 케이스
- 서버 response 300 -> BadResponse 반환 케이스

1Consumption commented on remy-movie/NetwokrPlatformTest/Helpers/MockURLProtocol.swift:
```diff
@@ -0,0 +1,44 @@
+
+final class MockURLProtocol: URLProtocol {
+    
+    static var requestHandler: ((URLRequest) throws -> (Data?, HTTPURLResponse))?
```
`HTTPURLResponse`가 아니라 `URLResponse`로 변경하면 `notHTTPURLResponse`의 경우도 테스트가 가능할듯 합니다.
결국 `HTTPURLResponse`도 `URLResponse`를 상속하고 있으니까요.
<img width="602" alt="image" src="https://user-images.githubusercontent.com/37682858/230454729-56cae2e6-420d-423a-bfdc-f433faca0a43.png">

yjjem replied:
생각도 못해본 점을 짚어주셔서 감사합니다 👍
MockURLProtocol이 requestHandler를 통해 URLResponse 타입을 받도록 수정했고 관련 테스트들을 추가했습니다.
> 517bed2bb9c2533f6f805074c84d23ee03253108

1Consumption commented on remy-movie/NetwokrPlatformTest/Helpers/MockURLProtocol.swift:
```diff
@@ -0,0 +1,44 @@
+            let (data, response) = try requestHandler(request)
+            
+            client?.urlProtocol(self, didReceive: response, cacheStoragePolicy: .notAllowed)
+            if let data = data {
+                client?.urlProtocol(self, didLoad: data)
+            }
```
여기에 else 문을 추가하고, `client?.urlProtocol(_:didFailWithError:)`에 `emptyData` 에러를 넘기면 `emptyData`도 검증할 수 있겠습니다.

yjjem replied:
서버에서 빈 데이터가 날아온 상황은 검증할 수 있게되었습니다. 하지만 해당 에러는 completion의 error를 통해 반환됩니다.
따라서 현재 코드에서는 data의 옵셔널 바인딩간에 발생할 수 있는 emptyData 에러는 검증할 수 없습니다. 이 점이 아쉽지만 어쩔 수 없는 부분일까요 ?

```swift
// NetworkManager load 메서드 내부
...

guard let data = data else {
     completion(.failure(.emptyData))
     return
}

...
```

1Consumption replied:
앗 그러네요. 매개변수로 옵셔널 타입을 못넘기는군요... 별 다른 방법이 떠오르지 않습니다. 일단은 어쩔 수 없다고 생각합니다.

1Consumption commented on remy-movie/NetwokrPlatformTest/NetwokrPlatformTest.swift:
```diff
@@ -0,0 +1,228 @@
+//  Copyright (c) 2023 Jeremy All rights reserved.
+
+@testable
+import remy_movie
```
이정도는 가독성을 위해 개행을 하지 않아도 되지 않을까... 라고 생각합니다. 가독성을 위해 개행을 했는데 오히려 가독성이 떨어지는 느낌이랄까요.

yjjem replied:
감사합니다. 반영했습니다 👍

1Consumption commented on remy-movie/NetwokrPlatformTest/NetwokrPlatformTest.swift:
```diff
@@ -0,0 +1,228 @@
+
+final class NetwokrPlatformTest: XCTestCase {
+    
+    var sut: NetworkManager!
```
외부에서 몰라도 되는 변수 같습니다~

yjjem replied:
private으로 수정했습니다 👍

1Consumption commented on remy-movie/NetwokrPlatformTest/NetwokrPlatformTest.swift:
```diff
@@ -0,0 +1,228 @@
+        
+        // Arrange
+        
+        let dummyURL = try XCTUnwrap(URL(string: "test.com"))
+        let dummyResponse = HTTPURLResponse(
+            url: dummyURL,
+            statusCode: 300,
+            httpVersion: "2.0",
+            headerFields: nil
+        )
```
테스트 케이스를 검증하는데 필요하니 dummy는 아니라고 생각합니다. stub에 가깝다고 보이는데요,
https://martinfowler.com/articles/mocksArentStubs.html 이 아티클을 한번 읽어보는 것을 추천드립니다.

yjjem replied:
좋은 아티클 공유 감사합니다 :) 

아티클을 읽어보니 **Dummy**는 함수 파라매터 요구 사항을 충족을 위해 사용고 실직적으로 코드 로직내에서 사용되지않는 오브젝트입니다. dummyResponse의 경우 함수 호출에 대한 응답을 테스트하기 위한 것이기 떄문에 **Stub**에 가깝다는 것에 동의합니다.

아래와 같이 수정했습니다 

```swift
let url = try XCTUnwrap(URL(string: "test.com"))
let data = Data()
let stubResponse = URLResponse(
    url: url,
    mimeType: "text/plain",
    expectedContentLength: 0,
    textEncodingName: nil
)
```

1Consumption commented on remy-movie/NetwokrPlatformTest/NetwokrPlatformTest.swift:
```diff
@@ -0,0 +1,228 @@
+            headerFields: nil
+        )
+        
+        MockURLProtocol.requestHandler = { request in
+            return (nil, try XCTUnwrap(dummyResponse))
+        }
```
`XCTUnwrap` 이라는게 있는줄 처음 알았네요!
이건 개인 취향인데, 저는 테스트 코드에서는 강제 언래핑을 마음껏 합니다. 런타임에 테스트코드가 돌 일이 없고, 만약 테스트를 돌렸을 때 강제언래핑한 부분에서 죽는다면 애초에 잘못된 테스트 코드니까요.

yjjem replied:
감사합니다 👍

1Consumption commented on remy-movie/NetwokrPlatformTest/NetwokrPlatformTest.swift:
```diff
@@ -0,0 +1,228 @@
+        // Act and Assert
+        
+        let loadExpectation = expectation(description: "load")
+        let expectedCode = 300
```
이 code 변수를 `dummyResponse`에 넣으면 어떨까요? 만약 지금 `expectedCode`를 수정한다면 `dummyResponse`도 같이 수정을 해야하는데, 깜빡할 여지가 있어보여요.

yjjem replied:
확실히 아래와 같이 수정하는 것이 맞는 것 같습니다. 감사합니다 👍
```swift
// test
        
let url = try XCTUnwrap(URL(string: "test.com"))
let data = Data()
let expectedStatusCode = 300
let stubResponse = HTTPURLResponse(
    url: url,
    statusCode: expectedStatusCode,
    httpVersion: "2.0",
    headerFields: nil
)
```

1Consumption commented on remy-movie/NetwokrPlatformTest/NetwokrPlatformTest.swift:
```diff
@@ -0,0 +1,228 @@
+        
+        MockURLProtocol.requestHandler = { request in
+            return (nil, try XCTUnwrap(dummyResponse))
+            
```
불필요한 개행이 있네요.

yjjem replied:
수정했습니다 👍

1Consumption reviewed (commented):
고생하셨습니다. 
두가지정도 제안할게 있습니다.

## 1. 워닝 제거
<img width="277" alt="image" src="https://user-images.githubusercontent.com/37682858/230456635-41daf3c2-3e03-49b2-aa8a-7b674d7098e8.png">

해당 메소드들에 반환값을 사용하지 않을수도 있다는 의미인 `@discardableResult` 어노테이션을 추가하거나, 테스트 코드 내에서 아래와 같이 해결할 수 있겠습니다.
<img width="636" alt="image" src="https://user-images.githubusercontent.com/37682858/230457022-77ccdcc7-5072-4950-85bd-41c09caf0d2f.png">

## 2. 테스트 커버리지 시각화
1. edit scheme에 들어가서
<img width="446" alt="image" src="https://user-images.githubusercontent.com/37682858/230457161-386b2879-c8e6-4dab-80cf-552bec940c24.png">

2. Test - options에서 Code Coverage를 some target으로 변경해준다음
<img width="926" alt="image" src="https://user-images.githubusercontent.com/37682858/230457298-1f8f7e88-3407-4459-bb9b-0cd6fa33c103.png">

3. 호스트앱을 추가해주세요.
<img width="926" alt="image" src="https://user-images.githubusercontent.com/37682858/230457356-50975a97-e14b-49a9-a8f5-222c42f74888.png">

4. 그리고 Code Coverage 옵션을 켜주면
<img width="1053" alt="image" src="https://user-images.githubusercontent.com/37682858/230457431-fdbc79e5-2ea7-428d-8b5a-760ec296e74e.png">

5. 테스트 타겟에서 실행한 부분은 초록색으로 실행하지 않은 부분은 빨간색으로 표시됩니다.
<img width="885" alt="image" src="https://user-images.githubusercontent.com/37682858/230457610-7970450c-7657-4ba6-acd6-c69d62687177.png">

이렇게 하면 어느 부분이 테스트되지 않았는지 쉽게 확인 할 수 있습니다. 물론 모든 코드를 초록색으로 만들라는건 아닙니다. 테스트가 필요한 부분인데 테스트가 되지 않고 있는지를 확인하면 좋겠다는 의도에서 제안드립니다.

yjjem committed:
refactor(MockURLProtocol): 데이터가 없을 시 didFailWithEror를 통해 emptyData Error를 반환하도록 수정

yjjem committed:
test: 전반적인 테스트 네이밍 수정

rename dummy -> stub
change MockURLProtocol requestHandler HTTPURLResponse type -> URLResponse type

yjjem committed:
test: 워닝 메세지 제거

yjjem committed:
test: expectedStatusCode 공통 상수로 변경

yjjem committed:
test: notHTTPURLResponse 검증 테스트 추가

1Consumption reviewed (approved):
고생하셨습니다.
머지하셔도 좋습니다👍
